### PR TITLE
update conditions on error for delete operation

### PR DESCRIPTION
**Pull request by surajkota**
Description of changes:
- Update conditions on delete in case of error. Right now the errors from delete are not reflected to the user
  - I have added a check to treat all errors as recoverable if the resource is being deleted as I am not sure what is the user experience if the resource hits a terminal condition after its deleted. Please advise

Testing:
Manually introduced an error from sdkDelete and verified it gets updated in conditions

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: templates/pkg/resource/sdk.go.tpl [hunk 1]
```diff
@@ -212,7 +212,7 @@ func (rm *resourceManager) updateConditions (
 		}
 	}
 
-	if rm.terminalAWSError(err) || err ==  ackerr.SecretTypeNotSupported || err == ackerr.SecretNotFound {
+	if !r.IsBeingDeleted() && (rm.terminalAWSError(err) || err ==  ackerr.SecretTypeNotSupported || err == ackerr.SecretNotFound) {
```
Comment (jaypipes) [inline templates/pkg/resource/sdk.go.tpl:215]: I suppose this is fine. But, would it be safer to just let the Terminal Condition be set on the resource even if it's being deleted?

File: templates/pkg/resource/sdk.go.tpl [hunk 2]
```diff
@@ -224,7 +224,7 @@ func (rm *resourceManager) updateConditions (
 			errorMessage = err.Error()
 		} else {
 			awsErr, _ := ackerr.AWSError(err)
-			errorMessage = awsErr.Message()
+			errorMessage = awsErr.Code() + ": " + awsErr.Message()
```
Comment (jaypipes) [inline templates/pkg/resource/sdk.go.tpl:227]: unrelated change...

File: templates/pkg/resource/sdk.go.tpl [hunk 3]
```diff
@@ -247,7 +247,7 @@ func (rm *resourceManager) updateConditions (
 			awsErr, _ := ackerr.AWSError(err)
 			errorMessage := err.Error()
 			if awsErr != nil {
-				errorMessage = awsErr.Message()
+				errorMessage = awsErr.Code() + ": " + awsErr.Message()
```
Comment (jaypipes) [inline templates/pkg/resource/sdk.go.tpl:250]: unrelated change...

File: templates/pkg/resource/sdk.go.tpl (hunk 1 above)
Comment (brycahta) [reply templates/pkg/resource/sdk.go.tpl:215]: I don't think it's safe to set Terminal Condition here because if deletion encounters an error, then wouldn't the controller stop reconciling without deleting the finalizer?

**Comment by surajkota** [comment]
@jaypipes @brycahta thanks for review

I have moved the unrelated change to separate PR, please take a look

File: templates/pkg/resource/sdk.go.tpl (hunk 1 above)
Comment (surajkota) [reply templates/pkg/resource/sdk.go.tpl:215]: will this as-is for now, until we have another strategy. If a customer hits this(should be a corner case), the worse case is controller retires with back off

**Review by A-Hilaly — APPROVED** [review]

File: templates/pkg/resource/sdk.go.tpl (hunk 1 above)
Comment (jaypipes) [reply templates/pkg/resource/sdk.go.tpl:215]: > I don't think it's safe to set Terminal Condition here because if deletion encounters an error, then wouldn't the controller stop reconciling without deleting the finalizer?

Yes, and that is what the controller should do I think. If the delete operation failed on the backend AWS service API call, then the resource will not be deleted from the AWS service, so we should keep the resource around on the k8s side, but in show a terminal condition.

Comment (brycahta) [reply templates/pkg/resource/sdk.go.tpl:215]: > Yes, and that is what the controller should do I think. If the delete operation failed on the backend AWS service API call, then the resource will not be deleted from the AWS service, so we should keep the resource around on the k8s side, but in show a terminal condition.

Does this mean the only way to delete that resource (after resolving any errors) is for users to manually remove the finalizer via `kubectl edit` ? That seems weird cx or am I missing some details

**Review by jaypipes — APPROVED** [review]
:+1:

File: templates/pkg/resource/sdk.go.tpl (hunk 1 above)
Comment (jaypipes) [reply templates/pkg/resource/sdk.go.tpl]: > > Yes, and that is what the controller should do I think. If the delete operation failed on the backend AWS service API call, then the resource will not be deleted from the AWS service, so we should keep the resource around on the k8s side, but in show a terminal condition.
> 
> Does this mean the only way to delete that resource (after resolving any errors) is for users to manually remove the finalizer via `kubectl edit` ? That seems weird cx or am I missing some details

The correct user experience for manually removing a resource that is in a stuck state: using `kubectl delete --force`
